edb/server/connpool/pool.py

```diff
@@ -841,6 +842,31 @@ def _tick(self) -> None:
                     self._log_to_snapshot(
                         dbname=block.dbname, event='reset-quota')
 
+            if not was_starving and self._new_blocks_waitlist:
+                # Mode D assumes all connections are already in use or to be
+                # used, depending on their `release()` to schedule transfers.
+                # When just entering Mode D, there can be a special case when
+                # no further `release()` will be called because all acquired
+                # connections were returned to the pool before `_tick()` got a
+                # chance to set `self._is_starving`, while some other blocks
+                # are literally starving to death (blocked forever).
+                #
+                # This branch handles this particular case, by stealing
+                # connections from the idle blocks and try to free them into
+                # the starving blocks.
+
+                for block in list(self._blocks.values()):
+                    while block.count_conns() > block.count_waiters():
+                        if (conn := block.try_steal()) is None:
+                            # no more from this block
+                            break
+
+                        elif not self._maybe_free_conn(block, conn):
+                            # put back the last stolen connection if we
+                            # don't need to steal anymore
+                            block.release(conn)
+                            return
```

Comment:
Oh, I see, never mind, I think this is correct.

Reply:
Wait, I still don't understand why it is `return`

Reply:
Oh, because we aren't starving anymore?

Reply:
`_maybe_free_conn()` returns False if it cannot `_find_most_starving_block()`, which means all starving blocks are fed, so we can exit the loop early here by `return`.

But `_maybe_free_conn()` also checks `_should_free_conn()` which defeats the `return` here, so I should split `_maybe_free_conn()` and only call the second half here.

Reply:
Not sure if separating them is such a good idea (more code shuffle and overgeneralizing is also bad). IMO what would actually help is wiring docstrings for all these internal methods we use, so that you can make sense of code by reading them, instead of tracing the code (which is what I had to do)

Reply:
That said if you want to keep this refactoring in feel free to do so

Reply:
I end up calling `_should_free_conn()` explicitly, instead of having `_maybe_free_conn()` do multiple things (and renamed to `_maybe_free_into_starving_blocks()` for better readibility)